### PR TITLE
API injections in Procedures

### PR DESCRIPTION
**Pull request by jakewins**
GIven a procedure with

```
 @API
 public GraphDatabaseService db;
```

  we inject the `GraphDatabaseService` in the procedure.

### CODE REVIEW

**Comment by lutovich** [comment]
@jakewins I find term 'API injection' quite confusing. What does it mean to inject an application programming interface? 
Imo annotation `@API` is also quite confusing because it does not remind any well known annotations like `@Inject`, `@Resource` or `@Autowired`.

**Comment by pontusmelke** [comment]
@lutovich It is not meant as general dependency injection mechanism like `@Inject` or `@Autowired`, think of it more like how you can use  `@QueryParam("query")` on fields in JAX-RS. Using `@API` gives you an API (GDS) to use with Neo4j.

**Comment by thobe** [comment]
Am I the only one with a strong dislike for injecting things into fields?
A better design in my opinion would be to annotate a particular constructor for use with injection, then inject the API dependencies through that constructor instead.
Injecting dependencies through a constructor typically makes a class easier to unit test.

```
@Target(ElementType.CONSTRUCTOR)
@Retention(RetentionPolicy.RUNTIME)
public @interface BlessedConstructor {}
```

Usage:

```
public class MyProcedures {
    enum NodeType implements Label { Person }
    enum EdgeType implements RelationshipType { KNOWS, OWNS_PET }

    private final GraphDatabaseService graphDb;
    public @BlessedConstructor MyProcedure( GraphDatabaseService graphDb ) {
        this.graphDb = graphDb;
    }

    public @Procedure Stream<Friend> friends( String name ) {
        Node person = graphDb.findNode( NodeType.Person, "name", name );
        return StreamSupport.stream( person.getRelationships( EdgeType.KNOWS ).spliterator(), false )
                .map( edge -> new Friend( person, edge.getOtherNode( person ) ) );
    }

    public @Procedure Stream<Pet> petsOf( String name ) {
        Node person = graphDb.findNode( NodeType.Person, "name", name );
        return StreamSupport.stream( person.getRelationships( Direction.OUTGOING, EdgeType.OWNS_PET ).spliterator(), false )
                .map( edge -> new Pet( person, edge.getEndNode() ) );
    } 

    public static class Friend {
        public final String me, friend;
        public Friend( Node me, Node friend ) {
            this.me = (String)me.getProperty( "name" );
            this.friend = (String)friend.getProperty( "name" );
        }
    }
    public static class Pet {
        public final String owner, animal, name;
        public Pet( Node owner, Node pet ) {
            this.owner = (String)owner.getProperty( "name" );
            this.animal = (String)pet.getProperty( "animal" );
            this.name = (String)pet.getProperty( "name" );
        }
    }
}
```

**Comment by lutovich** [comment]
@pontusmelke the most weird thing for me is 'API injection' term and thus name of the `@API` annotation. Imo 'API injection' makes no sense.

+1 for constructor injection instead of field injection.

**Comment by jakewins** [comment]
@lutovich @pontusmelke I think there's a misunderstanding - `API` was meant to mean general injection, which component to inject then being determined by the field type. The thought was something like `give me this API`. I think this would be fine to change to `Inject` or `Component` or something like that.

The reason we're looking at field injection instead of constructor injection is because we want to "own" the fields in the class, forcing it to contain only injected things (and perhaps static fields). The reason for that is that we want state to be treated as a special thing, perhaps similar to how it's treated in JMH - constructor injection makes user code own state management for the class, which we do not want.

Owning the fields and having tight control on state also makes it much easier for us to re-use procedure instances in the future, since they become immutable. Leaving the door open to not having to re-do the dependency injection before each invocation of the procedure seemed useful.

There's nothing stopping users from doing the pattern you've listed above @thobe, by keeping two constructors:

```
public class MyProcedures {
    enum NodeType implements Label { Person }
    enum EdgeType implements RelationshipType { KNOWS, OWNS_PET }

    private @API GraphDatabaseService graphDb;

    public MyProcedure() { }

    public MyProcedure( GraphDatabaseService graphDb ) {
        this.graphDb = graphDb;
    }

    public @Procedure Stream<Friend> friends( String name ) {
        Node person = graphDb.findNode( NodeType.Person, "name", name );
        return StreamSupport.stream( person.getRelationships( EdgeType.KNOWS ).spliterator(), false )
                .map( edge -> new Friend( person, edge.getOtherNode( person ) ) );
    }

    public @Procedure Stream<Pet> petsOf( String name ) {
        Node person = graphDb.findNode( NodeType.Person, "name", name );
        return StreamSupport.stream( person.getRelationships( Direction.OUTGOING, EdgeType.OWNS_PET ).spliterator(), false )
                .map( edge -> new Pet( person, edge.getEndNode() ) );
    } 

    public static class Friend {
        public final String me, friend;
        public Friend( Node me, Node friend ) {
            this.me = (String)me.getProperty( "name" );
            this.friend = (String)friend.getProperty( "name" );
        }
    }
    public static class Pet {
        public final String owner, animal, name;
        public Pet( Node owner, Node pet ) {
            this.owner = (String)owner.getProperty( "name" );
            this.animal = (String)pet.getProperty( "animal" );
            this.name = (String)pet.getProperty( "name" );
        }
    }
}
```

**Comment by thobe** [comment]
@jakewins except that means the field is no longer final, and you've cluttered up the class with a constructor that doesn't construct a well formed object.

**Comment by jakewins** [comment]
@thobe definitely true. To me, the value of that is lower than the value of forcing state to be a special snowflake, since that's been one of the main sources of instability in unmanaged extensions. I don't know that there's a way to agree on either side, it's rather subjective

**Comment by pontusmelke** [comment]
Oh in that case I agree with @lutovich  I don't like `@API`. 

Could we have `@Target(ElementType.CONSTRUCTOR, ElementType.FIELD)` and allow both?

**Comment by jakewins** [comment]
@pontusmelke agree on renaming API, disagree on allowing both. The intention of field injection is to also disallow any non-static member field on the procedure class that isn't injected, which mean we guarantee that procedure classes are immutable, which gives several nice benefits.

If we do want to allow state on the procedure classes, then constructor injection is much preferable (for the reasons @thobe listed) and we should drop field injection entirely. However, given how much pain state management has been in UEs, I think it's sensible to take a careful approach here.

**Comment by chrisvest** [comment]
@jakewins @pontusmelke If you are going to reflectively write to fields, then please make sure that you _fail_ if the fields are `final`. Even though it is currently technically possible to overwrite final fields with reflection, I don't want us to rely on it at all.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/FieldInjections.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                    field.getName(), cls.getSimpleName() );
+        }
+
+        if( !Modifier.isPublic( field.getModifiers() ) || Modifier.isFinal( field.getModifiers() ) )
+        {
+            throw new ProcedureException(  Status.Procedure.FailedRegistration,
+                    "Field `%s` on `%s` must be non-final and public.",
+                    field.getName(), cls.getSimpleName() );
+
+        }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/FieldInjections.java:145]: Okay, cool.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/proc/FieldInjectionsTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        @Resource
+        public int parentField;
+    }
+
+    public static class ChildProcedure extends ParentProcedure
+    {
+        @Resource
+        public int childField;
+    }
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/proc/FieldInjectionsTest.java:126]: You're missing a test for the case where the field is final.

**Pull request merged**